### PR TITLE
[SPARK-44735][SQL] Add warning msg when inserting columns with the same name by row that don't match up

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
This PR add a warning msg when inserting columns name with the same name by row but order not matched. Tell user can use `INSERT INTO BY NAME` to reorder columns to match with table schema.
It will be like:
image

### Why are the changes needed?
Optimize user usage scenarios.

### Does this PR introduce _any_ user-facing change?
Yes, sometimes will show some warning.

### How was this patch tested?
Test in local

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 1]
```diff
@@ -403,11 +403,19 @@ object PreprocessTableInsertion extends ResolveInsertionBase {
       insert.query
     }
     val newQuery = try {
+      val byName = hasColumnList || insert.byName
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:406]: this is only for v1 table, does v1 table have the same problem?

**Review by holdenk — APPROVED** [review]
LGTM thank you for adding this functionality :)

If no one else has concerns I'll try and merge this week.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 2]
```diff
@@ -455,6 +457,19 @@ object TableOutputResolver {
...
+  def suitableForByNameCheck(
+      byName: Boolean,
+      expected: Seq[Attribute],
+      queryOutput: Seq[Attribute]): Unit = {
+    if (!byName && expected.size == queryOutput.size &&
+      expected.forall(e => queryOutput.exists(p => conf.resolver(p.name, e.name))) &&
+      expected.zip(queryOutput).exists(e => !conf.resolver(e._1.name, e._2.name))) {
+      logWarning("The query columns and the table columns have same names but different " +
+        "orders. You can use INSERT [INTO | OVERWRITE] BY NAME to reorder the query columns to " +
+        "align with the table columns.")
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:469]: Users always insert table columns doesn't match the query columns. For example, table `student(name: STRING, address: STRING)` and table `location(code: STRING, name: STRING)`. The `student.address` references `location.name`.
I am neutral on this warning message.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:469]: I know that. We can extend the example above to satisfy the condition you mentioned.

Comment (holdenk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:469]: Right but it's a warning message so if a user is doing it intentionally they can ignore the warning.

**Comment by holdenk** [comment]
Merged to the current branch so should be in 4 :)

**Comment by holdenk** [comment]
Hmm that's weird let me double check tonight.

**Comment by holdenk** [comment]
Good catch, idk why the original merge didn't stick but it's merged now (checked the git log etc.)

**Pull request closed without merging**